controllers/nstemplateset/space_roles_test.go

```diff
@@ -0,0 +1,210 @@
+			AssertThatRoleBinding(t, "oddity-appstudio", "user2-space-admin", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			AssertThatRole(t, "oddity-appstudio", "space-viewer", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			AssertThatRoleBinding(t, "oddity-appstudio", "user3-space-viewer", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			AssertThatRoleBinding(t, "oddity-appstudio", "user4-space-viewer", memberClient).
+				Exists(). // created
+				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
+				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
+			// verify that the `last-applied-space-roles` annotation was set on namespace
+			lastApplied, err := json.Marshal(nsTmplSet.Spec.SpaceRoles)
+			require.NoError(t, err)
+			AssertThatNamespace(t, "oddity-appstudio", memberClient.Client).
+				HasAnnotation(toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey, string(lastApplied))
+		})
+
+		t.Run("update roles", func(t *testing.T) {
+
+			t.Run("add admin user", func(t *testing.T) {
+				// given
+				nsTmplSet := newNSTmplSet(commontest.MemberOperatorNs, "oddity", "appstudio",
+					withSpaceRoles(map[string][]string{
+						"admin-appstudio-abcde11": {"user1", "user2"},
+					}))
+				ns := newNamespace(nsTmplSet.Spec.TierName, "oddity", "appstudio", // ns.name=oddity=appstudio
+					withTemplateRefUsingRevision("abcde10"), // starting with an older revision
+					withWorkspaceLabel("oddity"),
+				)
+				mgr, memberClient := prepareSpaceRolesManager(t, nsTmplSet, ns)
+				_, err := mgr.ensure(logger, nsTmplSet)
+				require.NoError(t, err)
+				// add `user3` in admin space roles
+				nsTmplSet.Spec.SpaceRoles[0].Usernames = append(nsTmplSet.Spec.SpaceRoles[0].Usernames, "user3")
+
+				// when
+				createdOrUpdated, err := mgr.ensure(logger, nsTmplSet)
+
+				// then
+				require.NoError(t, err)
+				assert.True(t, createdOrUpdated)
+				// verify that role bindings for `user1` and `user2`` still exist and thet one was added for `user3`
+				AssertThatRole(t, "oddity-appstudio", "space-admin", memberClient).Exists()              // unchanged
+				AssertThatRoleBinding(t, "oddity-appstudio", "user1-space-admin", memberClient).Exists() // unchanged
+				AssertThatRoleBinding(t, "oddity-appstudio", "user2-space-admin", memberClient).Exists() // unchanged
+				AssertThatRoleBinding(t, "oddity-appstudio", "user3-space-admin", memberClient).Exists() // created
+			})
+
+			t.Run("remove admin user", func(t *testing.T) {
```

Comment:
could you please verify also the case when one whole SpaceRole is removed with all usernames? Expecting that the `Role` would be deleted from the namespace.

Reply:
done in 789711c (as a subtest of the creation test, to avoid too much code duplication)